Fix thick controllers and refactor code

kushal-dreamspring opened the pull request:
- Add slot_view.rb
- Add invoice_view.rb
- Add view specs for slot_view.rb & invoice_view.rb
- Merge park, unpark and all_cars logic to slot_controller
- Merge invoice and all_invoices logic to Invoice_controller
- Reorganise specs in accordance to MVC model
- Add missing gem 'pg' for postgres support
- Add 'dotenv-rails' gem
- Use DB_URL environment variable in app.rb
- Move private methods from slot_controller.rb to model/slot.rb
- Move Invoice logic from invoice_controller.rb to model/invoice.rb
- Remove duration field from invoice model
- Use DEFAULT_LOT_SIZE in slot_controller
- Use Invoice.find_by_id in unpark car action

kushal-dreamspring committed:
Move output logic to views and reorganise controllers

- Add slot_view.rb
- Add invoice_view.rb
- Add view specs for slot_view.rb & invoice_view.rb
- Merge park, unpark and all_cars logic to slot_controller
- Merge invoice and all_invoices logic to Invoice_controller
- Reorganise specs in accordance to MVC model

kushal-dreamspring committed:
Add missing gem 'pg' for postgres support

kushal-dreamspring committed:
Move Database URL to environment variable

- Add 'dotenv-rails' gem
- Use DB_URL environment variable in app.rb

kushal-dreamspring committed:
Refactor Models and Controllers

- Move private methods from slot_controller.rb to model/slot.rb
- Move Invoice logic from invoice_controller.rb to model/invoice.rb

jinifrancis commented on controller/slot_controller.rb:
```diff
@@ -0,0 +1,48 @@
+
+require_relative '../model/slot'
+require_relative '../model/car'
+
```
Remove empty line

jinifrancis commented on controller/slot_controller.rb:
```diff
@@ -0,0 +1,48 @@
+
+# Controller Class for Slot
+class SlotController
+  def initialize(size = 10)
```
Use a constant for the size instead of using magic number

jinifrancis commented on model/slot.rb:
```diff
@@ -8,4 +9,21 @@ def validate
+
+  def self.unpark_car(slot)
+    invoice_id = Invoice.create(slot[:car_id], slot[:entry_time]).id
+    invoice = Invoice.where(id: invoice_id).join(
```
You can use Invoice.find_by_id method. You already have the id. The logic is duplicated here.

kushal-dreamspring replied:
I am actually using the 'sequel' gem instead of ActiveRecord..
```
Invoice.find_by_id( 1)
 => nil 
Invoice.where(id: 1)
 => #<Sequel::SQLite::Dataset: "SELECT * FROM `invoices` WHERE (`id` = 1)"> 
```
In sequel find_by_id returns the result of the SQL query instead of the query itself. So I cannot chain the join method to it

jinifrancis replied:
You have the same logic in Invoice class method `find_by_id(invoice_id)` - Why are you not able to use the method in the model? Why are you repeating the logic here?

jinifrancis reviewed (approved):
I have added a couple of comments. Make the changes and you can merge this PR.

kushal-dreamspring committed:
Refactor Code

- Add DEFAULT_LOT_SIZE constant to SLot Controller
- Remove extra line from SlotController

jinifrancis commented on controller/slot_controller.rb:
```diff
@@ -0,0 +1,49 @@
+# Controller Class for Slot
+class SlotController
+
+  DEFAULT_LOT_SIZE = 10
```
Why is this constant not used in the spec?

jinifrancis commented on spec/controller/slot_controller_spec.rb:
```diff
@@ -0,0 +1,82 @@
+  registration_number = 'UP32EA7196'
+
+  it 'should have 10 slots' do
+    expect(Slot.count).to eq(10)
```
Why is the constant not used here?

jinifrancis commented:
I mentioned saving 3 fields for entry_time, exit_time and duration in the database. But I don't see the updates on this PR.

kushal-dreamspring committed:
Remove duration field from invoice model

- Remove duration field from 2_create_invoices.rb
- Refactor model/invoice.rb
- Refactor  invoice_model_spec.rb
- Add invoice_duration logic to invoice_view.rb

kushal-dreamspring committed:
Use DEFAULT_LOT_SIZE in slot_controller_spec.rb

kushal-dreamspring committed:
Use Invoice.find_by_id in unpark car action

jinifrancis commented on controller/slot_controller.rb:
```diff
@@ -0,0 +1,49 @@
+  end
+
+  def park_car(registration_number)
+    car_id = (Car.where(registration_number:).first || Car.create(registration_number:))[:id]
```
Finding a car by registration number or creating a new car to get the car object can be methods on the car model instead of doing the actions in the controller. From the car object, id can be identified.

jinifrancis commented on controller/slot_controller.rb:
```diff
@@ -0,0 +1,49 @@
+
+  def park_car(registration_number)
+    car_id = (Car.where(registration_number:).first || Car.create(registration_number:))[:id]
+    slot = Slot.where(car_id: nil).order(:id).first
```
find_empty_slot - can be method in the slot model which can be invoked here

jinifrancis commented on controller/slot_controller.rb:
```diff
@@ -0,0 +1,49 @@
+
+    raise 'Parking Lot is Full!!' unless slot
+
+    slot.update(car_id:, entry_time: Time.now).save
```
park_car_in_slot can be a method in slot model. Or the other way to go about this is `park_car` method in slot model, which takes care of finding an empty slot and park the car in that slot position.

jinifrancis reviewed (changes requested):
I have a couple more suggestions to make this set of code better in the thick model thin controller concept. Go through them and let me know when the changes are implemented. Feel free to raise concerns or questions if you have any.

kushal-dreamspring committed:
Refactor park_car action in Slot Controller

- Add get_or_create action in Car model
- Add park_car_in_slot action in Slot model
- Add find_empty_slot action in Slot model

jinifrancis reviewed (changes requested):
Hey Kushal! You have missed updating/adding model specs for the new methods on the model. Add corresponding specs to the new methods.

kushal-dreamspring committed:
Fix minor bugs and add model specs

- Convert unpark class method in slot model to instance method
- Add specs for new methods in car_model_spec.rb, invoice_model_spec.rb and slot_model_spec.rb

jinifrancis reviewed (approved):
You can merge this PR 🎉
